src/System.ComponentModel.Annotations/tests/System/ComponentModel/DataAnnotations/AssociatedMetadataTypeTypeDescriptionProviderTests.cs

```diff
@@ -9,33 +9,361 @@ namespace System.ComponentModel.DataAnnotations.Tests
     public class AssociatedMetadataTypeTypeDescriptionProviderTests
     {
         [Fact]
-        public static void Constructor_NullType_ThrowsArgumentNullException()
+        public void Ctor_NullType_ThrowsArgumentNullException()
         {
             AssertExtensions.Throws<ArgumentNullException>("type", () => new AssociatedMetadataTypeTypeDescriptionProvider(null));
         }
 
         [Fact]
-        public static void Constructor_NullAssociatedMetadataType_ThrowsArgumentNullException()
+        public void Ctor_NullAssociatedMetadataType_ThrowsArgumentNullException()
         {
             AssertExtensions.Throws<ArgumentNullException>("associatedMetadataType", () => new AssociatedMetadataTypeTypeDescriptionProvider(typeof(string), null));
         }
 
         [Fact]
-        public static void Validate_GetTypeDescriptor_ReturnsTypeDescriptor_ForTypeWithCustomDescription()
+        public void GetTypeDescriptor_MetadataHasFieldsNotPresentOnClass_ThrowsInvalidOperationException()
         {
-            var provider = new AssociatedMetadataTypeTypeDescriptionProvider(typeof(SomeClassWithMetadata));
-            ICustomTypeDescriptor typeDescriptor = provider.GetTypeDescriptor(typeof(SomeClassWithMetadata), null);
-            PropertyDescriptorCollection props = typeDescriptor.GetProperties();
-            PropertyDescriptor firstNameProp = props[nameof(SomeClassWithMetadata.FirstName)];
-            string displayName = firstNameProp.DisplayName;
+            var provider = new AssociatedMetadataTypeTypeDescriptionProvider(typeof(ClassWithInvalidMetadata));
+            Assert.Throws<InvalidOperationException>(() => provider.GetTypeDescriptor(typeof(ClassWithInvalidMetadata), null));
+        }
+
+        [Fact]
+        public void GetTypeDescriptorGetAttributes_NoAssociatedMetadataTypeWithoutMetadataTypeAttribute_ReturnsExpected()
+        {
+            void PerformTest(AssociatedMetadataTypeTypeDescriptionProvider provider)
+            {
+                ICustomTypeDescriptor typeDescriptor = provider.GetTypeDescriptor(typeof(ClassWithMetadata), null);
+                AttributeCollection attributes = typeDescriptor.GetAttributes();
+                Assert.Equal(1, attributes.Count);
+                Assert.Equal("typeName", Assert.IsType<TypeConverterAttribute>(attributes[typeof(TypeConverterAttribute)]).ConverterTypeName);
+            }
+
+            // Perform multiple times to test caching behaviour.
+            PerformTest(new AssociatedMetadataTypeTypeDescriptionProvider(typeof(ClassWithMetadata)));
+            PerformTest(new AssociatedMetadataTypeTypeDescriptionProvider(typeof(ClassWithMetadata)));
```

Comment:
It could also have just been a loop `for (int i = 0; i < 2; i++) // perform multiple times to test caching behavior`, which is what we do in other places.

Reply:
Also, even if we stick with the local function approach, I'd prefer to see this as a loop around the local function call site rather than duplicating the same line twice.  It makes it very clear that the same thing is being done twice, and can easily be increased or decreased in iteration count, whereas with this multiple call site approach one needs to spend time parsing both lines to see that they're identical.

Reply:
I've moved to a loop as it makes sense.